File: JAVA/lld-examples/medium/online-shopping-system/src/main/java/com/systemdesign/lld/shopping/OrderManager.java
package com.systemdesign.lld.shopping;

import java.time.LocalDateTime;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

/**
 * OrderManager handles order processing and management
 */
public class OrderManager {
    private Map<String, Order> orders;
    private Map<String, List<Order>> userOrders;
    private AtomicInteger orderCounter;
    
    public OrderManager() {
        this.orders = new ConcurrentHashMap<>();
        this.userOrders = new ConcurrentHashMap<>();
        this.orderCounter = new AtomicInteger(1);
    }
    
    /**
     * Create a new order from shopping cart
     */
    public Order createOrder(User user, ShoppingCart cart, PaymentMethod paymentMethod) {
        if (cart.isEmpty()) {
            return null;
        }
        
        // Validate cart items availability
        if (!cart.validateCart()) {
            return null;
        }
        
        String orderId = "ORDER-" + String.format("%06d", orderCounter.getAndIncrement());
        Order order = new Order(orderId, user.getUserId(), cart.getAllItems(), paymentMethod);
        
        // Reserve stock for ordered items
        for (CartItem item : cart.getAllItems()) {
            Product product = item.getProduct();
            product.reduceStock(item.getQuantity());
        }
        
        orders.put(orderId, order);
        userOrders.computeIfAbsent(user.getUserId(), k -> new ArrayList<>()).add(order);
        
        // Clear cart after order creation
        cart.clearCart();
        
        return order;
    }
    
    /**
     * Get order by ID
     */
    public Order getOrder(String orderId) {
        return orders.get(orderId);
    }
    
    /**
     * Get all orders for a user
     */
    public List<Order> getUserOrders(String userId) {
        return userOrders.getOrDefault(userId, new ArrayList<>());
    }
    
    /**
     * Update order status
     */
    public boolean updateOrderStatus(String orderId, OrderStatus status) {
        Order order = orders.get(orderId);
        if (order != null) {
            order.updateStatus(status);
            return true;
        }
        return false;
    }
    
    /**
     * Cancel order
     */
    public boolean cancelOrder(String orderId) {
        Order order = orders.get(orderId);
        if (order != null && order.getStatus() == OrderStatus.PENDING) {
            order.updateStatus(OrderStatus.CANCELLED);
            
            // Restore stock for cancelled items
            for (OrderItem item : order.getItems()) {
                Product product = item.getProduct();
                product.addStock(item.getQuantity());
            }
            
            return true;
        }
        return false;
    }
    
    /**
     * Get orders by status
     */
    public List<Order> getOrdersByStatus(OrderStatus status) {
        return orders.values().stream()
                .filter(order -> order.getStatus() == status)
                .collect(Collectors.toList());
    }
    
    /**
     * Get orders within date range
     */
    public List<Order> getOrdersByDateRange(LocalDateTime startDate, LocalDateTime endDate) {
        return orders.values().stream()
                .filter(order -> {
                    LocalDateTime orderDate = order.getOrderDate();
                    return orderDate.isAfter(startDate) && orderDate.isBefore(endDate);
                })
                .collect(Collectors.toList());
    }
    
    /**
     * Get total sales amount
     */
    public double getTotalSales() {
        return orders.values().stream()
                .filter(order -> order.getStatus() == OrderStatus.DELIVERED)
                .mapToDouble(Order::getTotalAmount)
                .sum();
    }
    
    /**
     * Get order statistics
     */
    public Map<String, Object> getOrderStats() {
        Map<String, Object> stats = new HashMap<>();
        stats.put("totalOrders", orders.size());
        stats.put("totalSales", getTotalSales());
        
        Map<OrderStatus, Long> statusCount = orders.values().stream()
                .collect(Collectors.groupingBy(Order::getStatus, Collectors.counting()));
        
        stats.put("ordersByStatus", statusCount);
        
        return stats;
    }
}

/**
 * Order represents a customer order
 */
class Order {
    private String orderId;
    private String userId;
    private List<OrderItem> items;
    private OrderStatus status;
    private PaymentMethod paymentMethod;
    private LocalDateTime orderDate;
    private double totalAmount;
    private String shippingAddress;
    
    public Order(String orderId, String userId, List<CartItem> cartItems, PaymentMethod paymentMethod) {
        this.orderId = orderId;
        this.userId = userId;
        this.paymentMethod = paymentMethod;
        this.status = OrderStatus.PENDING;
        this.orderDate = LocalDateTime.now();
        
        // Convert cart items to order items
        this.items = cartItems.stream()
                .map(cartItem -> new OrderItem(cartItem.getProduct(), cartItem.getQuantity()))
                .collect(Collectors.toList());
        
        // Calculate total amount
        this.totalAmount = items.stream()
                .mapToDouble(OrderItem::getTotalPrice)
                .sum();
    }
    
    public String getOrderId() { return orderId; }
    public String getUserId() { return userId; }
    public List<OrderItem> getItems() { return new ArrayList<>(items); }
    public OrderStatus getStatus() { return status; }
    public PaymentMethod getPaymentMethod() { return paymentMethod; }
    public LocalDateTime getOrderDate() { return orderDate; }
    public double getTotalAmount() { return totalAmount; }
    public String getShippingAddress() { return shippingAddress; }
    
    public void updateStatus(OrderStatus status) {
        this.status = status;
    }
    
    public void setShippingAddress(String address) {
        this.shippingAddress = address;
    }
    
    @Override
    public String toString() {
        return "Order{" +
                "orderId='" + orderId + '\'' +
                ", status=" + status +
                ", totalAmount=" + totalAmount +
                ", orderDate=" + orderDate +
                '}';
    }
}

/**
 * OrderItem represents an item in an order
 */
class OrderItem {
    private Product product;
    private int quantity;
    private double priceAtOrder;
    
    public OrderItem(Product product, int quantity) {
        this.product = product;
        this.quantity = quantity;
        this.priceAtOrder = product.getPrice(); // Store price at time of order
    }
    
    public Product getProduct() { return product; }
    public int getQuantity() { return quantity; }
    public double getPriceAtOrder() { return priceAtOrder; }
    
    public double getTotalPrice() {
        return priceAtOrder * quantity;
    }
    
    @Override
    public String toString() {
        return "OrderItem{" +
                "product=" + product.getName() +
                ", quantity=" + quantity +
                ", priceAtOrder=" + priceAtOrder +
                '}';
    }
}

/**
 * OrderStatus enum for order states
 */
enum OrderStatus {
    PENDING,
    CONFIRMED,
    PROCESSING,
    SHIPPED,
    DELIVERED,
    CANCELLED,
    RETURNED
} 

File: Python/lld-examples/medium/online-shopping-system/order_manager.py
from typing import List, Dict, Optional
from order import Order, OrderItem
from user import User
from shopping_cart import ShoppingCart
from payment import PaymentMethod

class OrderManager:
    """Order manager class for handling orders"""
    
    def __init__(self):
        self.orders: Dict[str, Order] = {}
        self.orders_by_user: Dict[str, List[Order]] = {}
    
    def create_order(self, user: User, cart: ShoppingCart, payment_method: PaymentMethod) -> Optional[Order]:
        """Create order from cart"""
        if cart.is_empty():
            return None
        
        # Create order items
        order_items = []
        for cart_item in cart.get_items():
            product = cart_item.product
            
            # Check stock availability
            if not product.is_in_stock(cart_item.quantity):
                return None
            
            # Reduce stock
            product.reduce_stock(cart_item.quantity)
            
            # Create order item
            order_item = OrderItem(product, cart_item.quantity, product.price)
            order_items.append(order_item)
        
        # Create order
        order = Order(user.user_id, order_items, payment_method)
        
        # Store order
        self.orders[order.order_id] = order
        
        if user.user_id not in self.orders_by_user:
            self.orders_by_user[user.user_id] = []
        self.orders_by_user[user.user_id].append(order)
        
        # Add to user's order history
        user.add_order(order.order_id)
        
        return order
    
    def get_order_by_id(self, order_id: str) -> Optional[Order]:
        """Get order by order ID"""
        return self.orders.get(order_id)
    
    def get_orders_by_user(self, user: User) -> List[Order]:
        """Get orders by user"""
        return self.orders_by_user.get(user.user_id, [])
    
    def update_order_status(self, order_id: str, status: str) -> bool:
        """Update order status"""
        order = self.get_order_by_id(order_id)
        if order:
            # This is simplified - in real implementation, you'd have proper status validation
            return True
        return False
    
    def cancel_order(self, order_id: str) -> bool:
        """Cancel order"""
        order = self.get_order_by_id(order_id)
        if order:
            order.cancel_order()
            
            # Restore stock
            for item in order.items:
                item.product.increase_stock(item.quantity)
            
            return True
        return False
    
    def get_all_orders(self) -> List[Order]:
        """Get all orders"""
        return list(self.orders.values())
    
    def get_order_count(self) -> int:
        """Get total order count"""
        return len(self.orders) 